common/modules/reserve_queue/src/reserve_queue.rs

```diff
@@ -0,0 +1,417 @@
+        nonce: Nonce,
+        amount: &Self::BigUint,
+        precision: u64,
+        queues: &mut Vec<u64>,
+    ) -> SCResult<()> {
+        let is_fungible = nonce == 0;
+
+        if is_fungible {
+            self.push_single_fungible(token_id, nonce, amount, precision, queues)
+        } else {
+            self.push_single_non_fungible(token_id, nonce, amount, precision, queues)
+        }
+    }
+
+    fn push_single_fungible(
+        &self,
+        token_id: &TokenIdentifier,
+        nonce: Nonce,
+        amount: &Self::BigUint,
+        _precision: u64,
+        _queues: &mut Vec<u64>,
+    ) -> SCResult<()> {
+        let exists = !self.reserve(token_id, nonce).is_empty();
+
+        if exists {
+            self.reserve(token_id, nonce).update(|x| x.amount += amount);
+        } else {
+            self.reserve(token_id, nonce)
+                .set(&ReserveNode::from(amount.clone(), 0));
+        }
+
+        Ok(())
+    }
+
+    fn push_single_non_fungible(
+        &self,
+        token_id: &TokenIdentifier,
+        nonce: Nonce,
+        amount: &Self::BigUint,
+        precision: u64,
+        queues: &mut Vec<u64>,
+    ) -> SCResult<()> {
+        require!(nonce != 0, "Nonce cannot be zero");
+
+        if self.reserve(token_id, nonce).is_empty() {
+            self.push_single_non_fungible_new(
+                token_id,
+                nonce,
+                amount,
+                precision,
+                queues,
+                Option::None,
+            );
+        } else {
+            self.push_single_non_fungible_existing(token_id, nonce, amount, precision, queues);
+        }
+
+        Ok(())
+    }
+
+    fn should_change_queue(&self, current: u64, new: u64) -> bool {
+        current != new
+    }
+
+    fn get_precision_or_default(&self, token_id: &TokenIdentifier) -> u64 {
+        let precision = self.precision(token_id).get();
+
+        if precision != 0 {
+            precision
+        } else {
+            DEFAULT_PRECISION
+        }
+    }
+
+    fn get_compressed_pairs(
+        &self,
+        pairs: &[NonceAmountPair<Self::BigUint>],
+    ) -> Vec<NonceAmountPair<Self::BigUint>> {
+        let mut pairs_compressed = Vec::<NonceAmountPair<Self::BigUint>>::new();
+
+        pairs.iter().filter(|&x| x.amount != 0).for_each(|pair| {
+            match pairs_compressed.iter().position(|x| x.nonce == pair.nonce) {
+                Some(index) => pairs_compressed[index].amount += &pair.amount,
+                None => pairs_compressed.push(pair.clone()),
+            }
+        });
+
+        pairs_compressed
+    }
+
+    fn get_queue_id(&self, amount: &Self::BigUint, precision: u64) -> u64 {
+        let mut queue_id = 0;
+
+        let precision_biguint = Self::BigUint::from(precision);
+        let mut amount_clone = amount.clone();
+        while amount_clone != 0 {
+            amount_clone /= &precision_biguint;
+            queue_id += 1;
+        }
+
+        queue_id
```

Comment:
Is this the binary search ? which sets the actual position for one token-nonce pair ?

Reply:
This just maps an amount (which can be whatever) to a bucket (a queue). Element with X amount will be added to a queue with respect to a configured precision